Evaluate argument of (DYNAMIC_)SECTION, TEST_CASE in static-analysis mode

mjerabek opened the pull request:
## Description

As the title says, it's fairly straightforward.

And also add configuration for `clang-tidy`, run it in CI, and fix/suppress the resulting warnings.

## GitHub Issues

Closes #2816

mjerabek committed:
Evaluate argument of (DYNAMIC_)SECTION in static-analysis mode

mjerabek commented:
Not sure if/where should I add tests for this.

mjerabek committed:
Evaluate argument of TEST_CASE in static-analysis mode

horenmar commented:
I am _reasonably sure_ that this breaks `SECTION` macros in some cases, so a test would be nice 😛 .

This is a valid `SECTION` start: `SECTION("short", "long description with many many words")`, and unless I am very mistaken, your change would break it.

mjerabek commented:
Huh, interesting. I had no idea it can be used like that :smile: 

Any tips where the tests for this should be?

mjerabek committed:
ci: add clang-tidy run

mjerabek commented:
- fixed the implementation (yes, the tests uncovered other bugs)
- added basic clang-tidy run to CI

Some limitations:
- Some files are checked multiple times, because they are present in `compile_commands.json` multiple times with different compile options. Also I'm quite certain that `run-clang-tidy` doesn't handle this well and we run it multiple times with the same configuration.
- `run-clang-tidy` doesn't stop on error, so it always runs in full. If we used cmake's clang-tidy integration, we could rely on ninja. But cmake doesn't support running clang-tidy without also running the compiler, so it would either take longer, or we would have to fake a "null compiler" (I've done that in work, it works well, but it's not very nice). Or we could put together a short python script that generates a ninja build file and run that (I would probably slightly prefer that to the current solution, as we could also better deal with duplicate files in `compile_commands.json`).
- Only the default checks are enabled at the moment. We should probably explicitly define what we want to check and what we want to consider an error.
- If there are errors, the job fails (good). But we ignore any warnings. :shrug: 

Also, I might add some clang-tidy-specific tests later to test that `clang-tidy` understands that `REQUIRE(expr)` implies `expr` (and also implement it, because now it most probably cannot understand that). But that's for a follow-up PR.

horenmar commented on tests/SelfTest/UsageTests/Benchmark.tests.cpp:
```diff
@@ -97,7 +97,7 @@ TEST_CASE("Benchmark containers", "[!benchmark]") {
         int sum = 0;
         for (int i = 0; i < size; ++i)
-            sum += array[i];
+            sum += array[i]; // NOLINT(clang-analyzer-core.uninitialized.Assign)
```
Wouldn't a better fix be changing `int array[size]` into `int array[size]{}` above? The test below would still require the change in `BENCHMARK` to happen to pass, but it should shut up the analyzer.

mjerabek replied:
Done.

horenmar commented on tests/ExtraTests/X91-AmalgamatedCatch.cpp:
```diff
@@ -14,7 +14,7 @@
-    auto i = GENERATE(1, 2, 3);
+    [[maybe_unused]] auto i = GENERATE(1, 2, 3);
     SECTION("a") {
         REQUIRE(1 != 4);
```
I _think_ this might've been supposed to be `REQUIRE(i != 4)`, and similarly in the next section, so the `maybe_unused` wouldn't be needed.

mjerabek replied:
Done.

horenmar commented on .github/workflows/linux-other-builds.yml:
```diff
@@ -103,3 +103,48 @@ jobs:
+          # Extras and examples with clang-tidy-10
+          - version: "10"
+            tidy_options: ''
+            build_description: Extras + Examples
```
```suggestion
            build_description: clang-tidy
```
instead?

mjerabek replied:
"clang-tidy" is already included in the job description. I've replaced this by "all", but it's not really _completely_ all. I guess it's hard to put variant-specific description when there is just one variant.

horenmar replied:
Right, then "extras + examples" is fine.

horenmar commented:
> run-clang-tidy doesn't stop on error, so it always runs in full. If we used cmake's clang-tidy integration, we could rely on ninja. But cmake doesn't support running clang-tidy without also running the compiler, so it would either take longer, or we would have to fake a "null compiler" (I've done that in work, it works well, but it's not very nice). Or we could put together a short python script that generates a ninja build file and run that (I would probably slightly prefer that to the current solution, as we could also better deal with duplicate files in compile_commands.json).

That's not great. Maybe we could merge the clang-tidy check into an already existing build? This would create a build with outsized build time, but we can't build all the builds in parallel either way, so as long as it gets scheduled reasonably early it shouldn't have much impact.

> Some files are checked multiple times, because they are present in compile_commands.json multiple times with different compile options. Also I'm quite certain that run-clang-tidy doesn't handle this well and we run it multiple times with the same configuration.

Yes, they are indeed compiled multiple times in the full test build. Some tests require building against custom configuration of the implementation binary, so they have to rebuild it. You can find them quickly by looking for targets that link against `Catch2_buildall_interface`.

> Only the default checks are enabled at the moment. We should probably explicitly define what we want to check and what we want to consider an error.

Definitely. The checks in clang-tidy can be a mess, with overlapping checks with different implementation, this came up in the original discussion for the static-analysis mode. Having a reasonable set that we can point to as being supported would be nice.

> Also, I might add some clang-tidy-specific tests later to test that clang-tidy understands that REQUIRE(expr) implies expr (and also implement it, because now it most probably cannot understand that).

It almost definitely cannot understand it. The reasoning there is only slightly simpler than figuring out `SECTION`s, and the analyzer helper mode has to help with those a lot. However, from cursory check, there is a relatively reasonable way of [telling clang-tidy about custom assertions](https://clang-analyzer.llvm.org/faq.html#custom_assert). It would require making the `REQUIRE` macro also change its definition based on the defines though.

mjerabek committed:
less copies and allocations in replaceInPlace

mjerabek committed:
add .clang-tidy config

mjerabek commented:
In the end, I have rewritten the CI job to use `cmake` to run `clang-tidy`, with a fake compiler (turned out it's not messy when you're not adding PCHs and ccache into the mix, which I now realize are both useless for just clang-tidy run; oh well, I'm putting that on the endless list of pending CI refactorings).

The `clang-tidy` checks are now explicitly configured. The rationale is following: allow-list whole categories, deny-list particular checks from the categories. This way, when a new clang version adds new checks, they are easily discovered (if they print a warning) and may be acted upon - either fix the findings or disable the check if it is deemed not useful. YMMV

Werror is enabled for some concrete checks.

Unfortunately, we cannot enable `bugprone-use-after-move`. First, it cannot be made to understand our custom `unique_ptr` (https://github.com/llvm/llvm-project/issues/36901). Second, it [doesn't understand](https://clang.llvm.org/extra/clang-tidy/checks/bugprone/use-after-move.html) that `if (i == 1)` and `if (i == 2)` are exclusive, so it doesn't work with the `SECTION` hinting.

We also have to disable `bugprone-macro-repeated-side-effects` because if wrongly triggers for `CATCH_MOVE` and `CATCH_FORWARD`, although the second expansion of the macro argument is in unevaluated context (no issue for that yet AFAICT).

mjerabek committed:
Fix various useful clang-tidy warnings

* bugprone-branch-clone
* bugprone-copy-constructor-init
* bugprone-empty-catch
* bugprone-sizeof-expression
* bugprone-switch-missing-default-case
* bugprone-unused-local-non-trivial-variable
* clang-analyzer-core.uninitialized.Assign
* clang-analyzer-cplusplus.Move
* clang-analyzer-optin.cplusplus.VirtualCall
* modernize-loop-convert
* modernize-raw-string-literal
* modernize-use-equals-default
* modernize-use-override
* modernize-use-using
* performance-avoid-endl
* performance-inefficient-string-concatenation
* performance-inefficient-vector-operation
* performance-noexcept-move-constructor
* performance-unnecessary-value-param (and improve generator example)
* readability-duplicate-include
* readability-inconsistent-declaration-parameter-name
* readability-non-const-parameter
* readability-redundant-casting
* readability-redundant-member-init
* readability-redundant-smartptr-get
* readability-static-accessed-through-instance
* unused variable in amalgamted tests

mjerabek committed:
releaseCommon: fix SyntaxWarning: invalid escape sequence '\s'

mjerabek commented:
So, I have split each check to its separate commit as you wanted. It's up to you now to select which ones do you want to keep and which ones you want to suppress instead.

horenmar commented on tests/SelfTest/UsageTests/Misc.tests.cpp:
```diff
@@ -396,7 +396,7 @@ TEMPLATE_PRODUCT_TEST_CASE("Product with differing arities", "[template][product
 TEMPLATE_LIST_TEST_CASE("Template test case with test types specified inside std::tuple", "[template][list]", MyTypes)
 {
-    REQUIRE(sizeof(TestType) > 0);
+    REQUIRE(sizeof(TestType) > 0); // NOLINT(bugprone-sizeof-expression)
```
Is this due to the expression desugaring into `sizeof(sizeof(TestType) > 0)`?

The point here is to just have some assertion in the test case, so it can be replaced with something else instead... For this one `STATIC_REQUIRE(std::is_arithmetic<TestType>::value)` should be fine.

mjerabek replied:
I'm not entirely certain. I think it has to do something with the tested type being `int`.

> `suspicious comparison of 'sizeof(expr)' to a constant

Agree that it's better to replace.

mjerabek replied:
Done.

horenmar commented on tests/SelfTest/UsageTests/Misc.tests.cpp:
```diff
@@ -406,7 +406,7 @@ struct NonDefaultConstructibleType {
 TEMPLATE_LIST_TEST_CASE("Template test case with test types specified inside non-default-constructible std::tuple", "[template][list]", MyNonDefaultConstructibleTypes)
 {
-    REQUIRE(sizeof(TestType) > 0);
+    REQUIRE(sizeof(TestType) > 0); // NOLINT(bugprone-sizeof-expression)
```
Same here, but maybe `is_trivially_copyable`.

mjerabek replied:
Done.

horenmar commented on tests/SelfTest/UsageTests/Misc.tests.cpp:
```diff
@@ -421,7 +421,7 @@ struct NonCopyableAndNonMovableType {
 TEMPLATE_LIST_TEST_CASE("Template test case with test types specified inside non-copyable and non-movable std::tuple", "[template][list]", NonCopyableAndNonMovableTypes)
 {
-    REQUIRE(sizeof(TestType) > 0);
+    REQUIRE(sizeof(TestType) > 0); // NOLINT(bugprone-sizeof-expression)
```
`default_constructible` or something.

mjerabek replied:
Done.

horenmar commented on tests/SelfTest/UsageTests/ToStringGeneral.tests.cpp:
```diff
@@ -27,9 +27,8 @@ TEST_CASE( "Character pretty printing" ){
+        for (char c : chars){
+            REQUIRE(c == c);
         }
     }
```
These tests are all wrong, I'll fix them in main and you can drop the commit.

mjerabek replied:
Dropped.

horenmar commented on tests/SelfTest/UsageTests/Message.tests.cpp:
```diff
@@ -80,28 +80,28 @@ TEST_CASE( "Output from all sections is reported", "[failing][messages][.]" ) {
         std::clog << "writes";
         std::cerr << " to error";
-        std::clog << " streams" << std::endl;
+        std::clog << " streams\n";
```
I suspect these should all explicitly flush for 100% correctness.

mjerabek replied:
I also wondered about that, but the tests passed, so I thought it should be ok. We can either use `<< '\n' << std::flush` or disable the check.

horenmar replied:
Add the `std::flush`, that's what Catch2 implementation does when it relies on flush happening.

mjerabek replied:
Added `<< std::flush`.

horenmar commented on tests/SelfTest/UsageTests/Misc.tests.cpp:
```diff
@@ -158,9 +158,9 @@ TEST_CASE( "looped tests", "[.][failing]" ) {
-    std::clog << "A string sent to stderr via clog" << std::endl;
+    std::cout << "A string sent directly to stdout\n";
+    std::cerr << "A string sent directly to stderr\n";
+    std::clog << "A string sent to stderr via clog\n";
```
ditto

mjerabek replied:
Added `<< std::flush`.

horenmar commented on tests/SelfTest/UsageTests/Matchers.tests.cpp:
```diff
@@ -1081,6 +1081,7 @@ struct MatcherWasMovedOrCopied : std::exception {
 
 struct ThrowOnCopyOrMoveMatcher : Catch::Matchers::MatcherGenericBase {
     ThrowOnCopyOrMoveMatcher() = default;
+    // NOLINTNEXTLINE(bugprone-copy-constructor-init)
```
I am looking at the docs, and apparently changing the code below to `Catch::Matcher::MatcherGenericBase(other)` would work.

Or maybe deleting the initialization of the base class completely might work, as the argument is already marked as unused via deleted name.

mjerabek replied:
Deleting the call would trigger a different warning :smile: Calling the copy constructor _might_ work, though. I just find it all mightily stupid.

mjerabek replied:
Calling the copy constructor worked. I have removed the suppressions.

horenmar commented on tests/SelfTest/UsageTests/Matchers.tests.cpp:
```diff
@@ -1079,6 +1079,7 @@ struct MatcherWasMovedOrCopied : std::exception {
 struct ThrowOnCopyOrMoveMatcher : Catch::Matchers::MatcherGenericBase {
     ThrowOnCopyOrMoveMatcher() = default;
 
+    // NOLINTBEGIN(readability-redundant-member-init)
```
Same as the other lint here, is it still needed after a code cleanup?

mjerabek replied:
Same here.

horenmar commented on src/catch2/internal/catch_reporter_spec_parser.cpp:
```diff
@@ -117,7 +117,7 @@ namespace Catch {
             auto key = kv.key, value = kv.value;
 
-            if ( key.empty() || value.empty() ) {
+            if ( key.empty() || value.empty() ) { // NOLINT(bugprone-branch-clone)
```
What's supposed to be duplicated here?

mjerabek replied:
The `else` branch waaay down.

horenmar replied:
ugh, that's dumb

mjerabek replied:
Very :smile:
